docs: Aliases overview and domain model (#4705)
hashicorp/boundary: website/content/docs/concepts/aliases.mdx

---
layout: docs
page_title: Aliases
description: |-
  Aliases let you associate a string with a Boundary resource. Learn how to use an alias instead of a target ID when you connect.
---

# Aliases

An alias is a globally unique, DNS-like string that is associated with a destination resource.
Exampes of valid aliases include `webserver` and `webserver.boundary`.
You can establish a session to a target by referencing its alias, instead of having to provide a target ID or target name and scope ID.

> For example, if you have an alias `boundary.dev`, you can use it to connect to a target with the following command: `boundary connect ssh boundary.dev`.

Reviewer: Same here, would prefer to see this broken out into shell-session format

## Background

When you create a resource in Boundary, it assigns the resource a randomly generated identifier.
You must use those IDs to perform actions in Boundary.

> When you connect to a target using the terminal, you must reference the target ID or target name and scope name. As an example, to SSH to a target, you can type `boundary connect ssh -target-id ttcp_123456789`.

Reviewer: ```suggestion
When you connect to a target using the terminal, you must reference the target ID or target name and scope name. As an example, to SSH to a target, you can execute the command `boundary connect ssh -target-id ttcp_123456789`.
```

Aliases simplify this process.
When you create an alias and associate it with a target, you can later use the alias `value` instead of the target ID in commands.
Boundary automatically resolves to the target that the alias references.

## Permissions

The existence of an alias for a Boundary resource does not change how permissions function.
Anyone can attempt to use an alias to access a target, but if you do not have permission to access the target, the attempt fails.
You can create an alias for a target, even if you do not have permission to access the target.

Separating the permissions from aliases and destination resources means a different set of people can manage the aliases than the people who have permission to operate on targets.
For example, you may have a project with a sensitive set of targets.
You can configure Boundary to allow a select few users to manage those targets, while a different group of users manage the aliases.

## Naming conventions

An alias is a globally unique, DNS-like string that is associated with a destination resource.
The alias `value` parameter does not have to be delimited by a suffix, and can be just a hostname.

> Exampes of valid aliases include `webserver` and `webserver.boundary`.

Reviewer: This sentence was repeated at the top of the page. Are there other valid examples or naming formats we want to call out here? Or are these only valid formats:

`value1`
`value1.value2.valueX`

## Scopes

At this time, you can only create aliases in the `global` scope.

> An alias' destination target can be in any scope, however.

Reviewer: I'd suggest removing this possessive for accessibility 

```suggestion
The destination target for an alias can be in any scope, however.
```

If you delete a project, Boundary clears the `destination_id` parameter for any aliases that resolve to targets in that project, so that they no longer function.

## Connect to a target using an alias

Whenever you could use the `-id` flag or `-target` flag in the CLI, you can substitute an alias.

For example, you can use the following command to connect to an SSH target with the ID `ttcp_1234567890`:

```shell-session

> boundary connect ssh -target -id ttcp_1234567890

Reviewer: ```suggestion
$ boundary connect ssh -target -id ttcp_1234567890
```

```

If you configured an alias named `example.alias.boundary` for the target, you could alternatively use the following command to connect to the target:

```shell-session

> boundary connect ssh example.alias.boundary

Reviewer: ```suggestion
$ boundary connect ssh example.alias.boundary
```

```

## Create aliases

There are multiple ways that you can create aliases and associate them with targets in Boundary.
You can:

- [Create a new alias for an existing target](/boundary/docs/concepts/aliases#create-a-new-alias-for-an-existing-target)
- [Create a new target and alias at the same time](/boundary/docs/concepts/aliases#create-a-new-target-and-alias)
- [Associate an existing alias with a target](/boundary/docs/concepts/aliases#associate-an-existing-alias-with-a-target)

> ### Create a new alias for an existing target

Reviewer: this feels more natural to me

```suggestion
### Create an alias for an existing target
```

   The alias `value` can be a hostname or a DNS-like string.
   - `-authorize-session-host-id=<string>` - Optionally indicates the host ID to use when you use the alias to authorize a session.

</Tab>

<Tab heading="UI">

1. Log in to Boundary.
1. Select **Aliases** in the navigation pane.
1. Click **New Alias**.
1. Complete the following fields:
   - **Name**: (Optional) Enter an optional name for the alias to use for identification purposes.
   - **Description**: (Optional) Enter an optional description for the alias to use for identification purposes.
   - **Type**: Select **Target**.
   At this time, targets are the only Boundary resources that supports aliasing.
   - **Alias Value**: Enter the string that you want to use as the alias to represent the target.
   An alias's value can be a hostname or a DNS-like string.
   - **Target ID**: (Optional) Specify any targets you want to associate with the alias.
   - **Host ID**: (Optional) Enter an optional host ID, if you would like to specify that the alias always uses the same host when you use it to connect to a target.
1. Click **Save**.

</Tab>

</Tabs>

> ### Create a new target and alias

Reviewer: I found this header a little confusing, just suggesting different verbage

```suggestion
### Create an alias during target creation
```

Author: I like this much better. Thanks!

---
layout: docs
page_title: Aliases
description: |-
  Aliases let you associate a string with a Boundary resource. Learn how to use an alias instead of a target ID when you connect.
---

# Aliases

An alias is a resource that lets you associate a string with a specific target.

> The string must be a globally unique value which follows DNS naming constraints.

Reviewer 2: Not necessarily true, as we support hostnames(e.g an alias of `webserver` as well as DNS-like strings `webserver.boundary`. May be worth rephrasing to something like: An alias is a globally unique, DNS-like string associated with a destination resource. Examples of valid aliases may be `webserver` or `webserver.boundary`.

You can use the alias instead of the target ID when you run a command, and Boundary automatically substitutes the target resource that the alias references.

## Background

When you create a resource in Boundary, it assigns the resource a randomly generated identifier.
You must use those IDs to perform actions in Boundary.

> For example, when you connect to a target using the clipboard, you must reference the target ID by typing `boundary ssh connect -target-id tccp_123456789`.

Reviewer 2: ``
```suggestion
When you connect to a target using the terminal, you must reference the target ID or target-name & scope-name. As an example, to SSH to a target, you can type `boundary connect ssh -target-id ttcp_123456789`.
```

## Permissions

The existence of an alias for a Boundary resource does not change how permissions function.
Anyone can attempt to use an alias to access a target, but if you do not have permission to access the target, the attempt fails.
You can create an alias for a target, even if you do not have permission to access the target.

Separating the permissions from aliases and destination resources means a different set of people can manage the aliases than the people who have permission to operate on targets.
For example, you may have a project with a sensitive set of targets.
You can configure Boundary to allow a select few users to manage those targets, while a different group of users manage the aliases.

## Naming conventions

Aliases must comply with DNS naming rules.
The alias `value` parameter must meet the following criteria:

- At least one label must be joined together by a period ( . )
- Consists of alphanumeric characters A - Z, 0 - 9, and hyphens ( - )
- Not case sensitive
- Cannot start or end with a hyphen ( - )
- Between 1 and 63 characters long
- The final label cannot be all numeric characters

> 

Reviewer 2: We should note that it does not have to be delimited by a suffix, and can just be a hostname. 

E.g the value can just be `webserver` instead of `webserver.boundary`

1. Log in to Boundary.
1. Use the following command to create an alias:

   ```shell-session
   $ boundary aliases create target \
       -description 'This is an example alias for target tcp_1234567890' \
       -destination-id tcp_1234567890
       -name Example Boundary alias \
       -scope-id global \
       -value example.alias.boundary \
       -authorize-session-host-id hst_1234567890 \
   ```

   You can use any of the following [attributes](/boundary/docs/concepts/domain-model/aliases) when you create an alias:

   - `-description=<string>` - Specifies the optional description you want to use for identification purposes.
   - `-destination-id=<string>` - Specifies the ID of the target that the alias references.
   - `-name=<string>` - Specifies the optional name you want to use to describe the alias for identification purposes.
   - `-scope-id=<string>` - Scope in which to create the alias. The default is `global`.
   You can also specify the scope using the BOUNDARY_SCOPE_ID environment variable.
   At this time, aliases are only supported for the global scope.
   - `-value=<string>` - Specifies the string that you want to use as the alias to represent the target.

>    The alias `value` must comply with DNS naming rules.

Reviewer 2: ```suggestion
   The alias `value` can be a hostname or a DNS-like string.
```

   - `-authorize-session-host-id=<string>` - Optionally indicates the host ID to use when you use the alias to authorize a session.

</Tab>

<Tab heading="UI">

1. Log in to Boundary.
1. Select **Aliases** in the navigation pane.
1. Click **New Alias**.
1. Complete the following fields:
   - **Name**: (Optional) Enter an optional name for the alias to use for identification purposes.
   - **Description**: (Optional) Enter an optional description for the alias to use for identification purposes.
   - **Type**: Select **Target**.
   At this time, targets are the only Boundary resources that supports aliasing.
   - **Alias Value**: Enter the string that you want to use as the alias to represent the target.

>    The alias value must comply with DNS naming rules.

Reviewer 2: ```suggestion
   An alias's value can be a hostname or a DNS-like string.
```

1. Click **New Target**.
1. Complete the following fields:
   - **Name**: Enter the target name for identification purposes.
   - **Description**: (Optional) Enter an optional description for identification purposes.
   - **Type**: Select the target type.
   You can create SSH or TCP targets.
   - **Target Address**: (Optional) Enter a valid IP address or DNS name.
   Alternatively, you can configure host catalogs and host sets.
   - **Default Port**: (Optional) Enter an optional default port for the target to use for connections.
   - **Default Client Port**: (Optional) Enter an optional local proxy port on which to listen when a session is started on a client.
   - **Maximum Duration**: (Optional) Enter an optional maximum duration for sessions on this target, in seconds.
   - **Maximum Connection**: (Optional) Enter the maximum number of connections allowed per session on this target.
   For unlimited connections, enter `-1`.
   - **Workers**: (Optional) Select whether you want the worker to function as an ingress and/or egress worker.
   - **Aliases**: (Optional) Enter the value fpr any aliases you want to associate with this target, and then click **Add**.

>    The value must match DNS naming rules.

Reviewer 2: ```suggestion
   An alias's value can be a hostname or a DNS-like string.
```

   You can also specify the scope using the BOUNDARY_SCOPE_ID environment variable.
   At this time, aliases are only supported for the global scope.
   - `-value=<string>` - Specifies the string that you want to use as the alias to represent the target.
   The alias `value` must comply with DNS naming rules.
   - `-authorize-session-host-id=<string>` - Optionally indicates the host ID to use when you use the alias to authorize a session.

</Tab>

<Tab heading="UI">

1. Log in to Boundary.
1. Select **Targets** in the navigation pane.
1. Select the target you want to add an alias to.
1. Under the **Aliases** heading in the right sidebar, click **Add an alias**.
1. Complete the following fields:
   - **Name**: (Optional) Enter an optional name for the alias to use for identification purposes.
   - **Description**: (Optional) Enter an optional description for the alias to use for identification purposes.
   - **Type**: Select **Target**.
   At this time, targets are the only Boundary resources that supports aliasing.
   - **Alias Value**: Enter the alias value you want to use in commands to represent the target.

>    It must comply with DNS naming rules.

Reviewer 2: ```suggestion
   An alias's value can be a hostname or a DNS-like string.
```
